[TaskManager] Add TaskManager doc to FF.com (#11981)
microsoft/FluidFramework: docs/content/docs/data-structures/task-manager.md

### Consensus Based DDS
An important note about TaskManager is that it is a consensus based DDS. This essentially means that ops are not accepted until every client acknowledge and accepts the op. This differs from an "optimistic" DDS (i.e. [SharedMap]({{< relref "./map.md" >}})) which immediately accept ops and then relays them to other clients. For more information regarding different types of DDSes, click [here]({{< relref "./overview.md" >}}).

## Usage

### APIs

The `TaskManager` object provides a number of methods to manage the execution of tasks. Please note: each API requires an input of `taskId` which is type `string`.

- `volunteerForTask(taskId)` -- Adds the client to the task queue **once**. It returns a promise that resolves `true` if the client is assigned the task and `false` if the task was completed by another client. It will throw an error if the client disconnects while in queue.

> - `subscribeToTask(taskId)` -- Will continuously add the client to the task queue. Does not return a value, and will therefore require lisnening to [events](#events) to determine if the task is assigned, lost, or completed.

Reviewer: We need an explanation of what kind of scenarios call for `volunteerForTask` and what kind call for `subscribeToTask`. And at least one concrete example of each. If not in this article, then somewhere else that is linked from here.

- `subscribed(taskId)` -- Returns a boolean to indicate if the client is subscribed to the task.
- `complete(taskId)` -- Will release all clients from the task queue, including the currently assigned client.
- `abandon(taskId)` -- Exits the queue and releasing the task if currently assigned. Will also unsubscribe from the task (if subscribed).
- `queued(taskId)` -- Returns a boolean to indicate if the client is in the task queue (being assigned a task is still considered queued).
- `assigned(taskId)` -- Returns a boolean to indicate if the client is assigned the task.

### Events

> `TaskManager` is an `EventEmitter`, and will emit events when a task is assigned to the client or released. Each of the below events fire with an event listener that contains a callback argument `taskId`. This represents the task for which the event was fired for.

Reviewer: ```suggestion
`TaskManager` is an `EventEmitter`, and will emit events when a task is assigned to the client or released. Each of the following events fires with an event listener that contains a callback argument `taskId`. This represents the task for which the event was fired.
```

- `assigned` -- Fires when the client reaches the top of the task queue and is assigned the task.

> - `lost` -- Fires when the client disconnects while being assigned a task and is removed from the queue.

Reviewer: This could use some clarification. "while being assigned a task" means "the process of assigning the task to the client is underway but not yet complete". I suspect that what's actually meant here is "after having been assigned the task" (that is, if the disconnection occurs while `assigned(taskId)` is true).

> - `completed` -- Fires when the assigned client successfully calls `complete()`.

Reviewer: Does this fire only on the client that completes the task or on all clients?

Author: All clients, this will be clarified in the next commit

### Creation

> To create a `TaskManager`, call the static create method:

Reviewer: Either link to a reference topic for the `create` method or explain what the two parameters are.

```typescript
const taskManager = TaskManager.create(this.runtime, id);
```

## Examples

### Basic Example -- `volunteerForTask()`

The following is a basic example for `volunteerForTask()`. Note that we check the `boolean` return value from the promise to ensure that the task was not completed by another client.

```typescript
const myTaskId = "myTaskId";

taskManager.volunteerForTask(myTaskId)
  .then((isAssigned: boolean) => {
    if (isAssigned) {
      console.log("Assigned task.");

      // We setup a listener in case we lose the task assignment while executing the code.
      const onLost = (taskId: string) => {
        if (taskId === myTaskId) {
          // The task assignment has been lost, therefore we should halt execution.
          stopExecutingTask();
        }
      };
      taskManager.on("lost", onLost);

      // Now that we are assigned the task we can begin executing the code.
      executeTask()
        .then(() => {
          // We should remember to turn off the listener once we are done with it.

>           taskManager.off("lost", onLost);

Reviewer: Shouldn't `complete` be called here too? If not, how is "complete" triggered?

Author: I think I wanted to imply that `complete()` is being called at the end of `executeTask()`, but that's definitely not clear and can be done inside the `then` block as well. I'll add a note here and inside the `subscribeToTask()` example as well.

```typescript
const myTaskId = "myTaskId";

const onAssigned = (taskId: string) => {
  console.log(`Client was assigned task: ${taskId}`);
  if (taskId === myTaskId) {
    // Now that we are assigned the task we can begin executing the code.
    executeTask();
  }
}

const onLost = (taskId: string) => {
  console.log(`Client released task: ${taskId}`);
    if (taskId === myTaskId) {
      // This client is no longer assigned the task, therefore we should halt execution.
      stopExecutingTask();
    }
}

const onCompleted = (taskId: string) => {
  console.log(`Task ${taskId} completed by another client`);
  if (taskId === myTaskId) {
    // Make sure we turn off the event listeners now that we are done with them.
    taskManager.off("assigned", onAssigned);
    taskManager.off("lost", onLost);
    taskManager.off("completed", onCompleted);
  }

}

taskManager.on("assigned", onAssigned);
taskManager.on("lost", onLost);

> taskManager.on("completed", onCompleted);

Reviewer: It's not clear how/where "completed" is triggered.

### Consensus Based DDS
An important note about TaskManager is that it is a consensus based DDS. This essentially means that operations are not accepted until every client acknowledge and accepts the operation. This differs from an "optimistic" DDS (i.e. [SharedMap]({{< relref "./map.md" >}})) which immediately accept ops and then relays them to other clients. For more information regarding different types of DDSes, click [here]({{< relref "./overview.md" >}}).

## Usage

### APIs

The `TaskManager` object provides a number of methods to manage the execution of tasks. Please note: each API requires an input of `taskId` which is type `string`.

- `volunteerForTask(taskId)` -- Adds the client to the task queue **once**. It returns a promise that resolves `true` if the client is assigned the task and `false` if the task was completed by another client. It will throw an error if the client disconnects while in queue.

> - `subscribeToTask(taskId)` -- Will continuously add the client to the task queue. Does not return a value, and will therefore require lisnening to [events](#events) to determine if the task is assigned, lost, or completed.

Reviewer: ```suggestion
- `subscribeToTask(taskId)` -- Will continuously add the client to the task queue. Does not return a value, and will therefore require listening to [events](#events) to determine if the task is assigned, lost, or completed.
```

- `subscribed(taskId)` -- Returns a boolean to indicate if the client is subscribed to the task.
- `complete(taskId)` -- Will release all clients from the task queue, including the currently assigned client.
- `abandon(taskId)` -- Exits the queue and releasing the task if currently assigned. Will also unsubscribe from the task (if subscribed).
- `queued(taskId)` -- Returns a boolean to indicate if the client is in the task queue (being assigned a task is still considered queued).
- `assigned(taskId)` -- Returns a boolean to indicate if the client is assigned the task.

### `volunteerForTask()` vs `subscribeToTask()`

Although both APIs are ultimately used to join the task queue, they have two key differences which impacts which should be used in any given scenario. The first key difference is that `volunteerForTask()` returns a `Promise`, while `subscribeToTask()` is synchronous and will rely on events. Second, `volunteerForTask()` will only enter the client into the task queue **once**, while `subscribeToTask()` will re-enter the client into the task queue if the client disconnects and later reconnects.

> Due to these differences, `volunteerForTask()` is better suited for one-time tasks such as data imports or migrations. For an example, see [the schema upgrade demo](#external-examples). On the other hand, `subscribeToTask()` is prefered for ongoing tasks that have no defintive end. For an example, see [the task selection demo](#external-examples).

Reviewer: ```suggestion
Due to these differences, `volunteerForTask()` is better suited for one-time tasks such as data imports or migrations. For an example, see [the schema upgrade demo](#external-examples). On the other hand, `subscribeToTask()` is prefered for ongoing tasks that have no definitive end. For an example, see [the task selection demo](#external-examples).
```
